apps/src/lib/node/ledger/ethereum_node/mod.rs

```diff
@@ -28,15 +28,51 @@ pub enum Error {
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-/// Run the Ethereum fullnode. If it stops or an abort
-/// signal is sent, this processes is halted.
-pub async fn run(
-    mut ethereum_node: EthereumNode,
+/// Monitor the Ethereum fullnode. If it stops or an abort
+/// signal is sent, the subprocess is halted.
+pub async fn monitor(
+    ethereum_node: Subprocess,
+    abort_recv: Receiver<Sender<()>>,
+) -> Result<()> {
+    match ethereum_node {
+        Subprocess::Mock(node) => monitor_mock(node, abort_recv).await,
+        Subprocess::Geth(node) => monitor_real(node, abort_recv).await,
+    }
+}
+
+async fn monitor_real(
```

Comment:
Could we implement a common trait for the nodes the has the `kill` and `wait` methods and then only have one `monitor` method instead of two?

Reply:
:+1: will try implementing an `[async_trait]` again, I tried initially and was defeated because of something to do with sending futures across threads in the real `EthereumNode::new`, but with just the `kill` and `wait` methods it should hopefully just work

Reply:
added a `Monitorable` trait which gets rid of some code duplication, though this code can still be simplified a lot further. When merging this PR, I can open an issue to make it so that `run_aux` does not need dummy node/oracle handles, since the dummy node/oracle are only needed afaict because `run_aux` needs a hardcoded five tasks to manage at the moment (shell, abci server, broadcaster, eth-fullnode, oracle)